service/matching/taskReader.go

```diff
@@ -421,3 +376,120 @@ func (tr *taskReader) newDispatchContext(isolationGroup string) (context.Context
+				tr.completeTask(taskInfo, nil)
+				return false, true
+			}
+			// it should never happen, unless there is a bug in 'getIsolationGroupForTask' method
+			tr.logger.Error("taskReader: unexpected error getting isolation group", tag.Error(err))
+			tr.completeTask(taskInfo, err)
+			return false, true
+		}
+
+		if group == isolationGroup {
+			// no change, retry to dispatch the task again
+			return false, false
+		}
+
+		// ensure the isolation group is configured and available
+		_, taskGroupReaderIsPresent := tr.taskBuffers[group]
+		if !taskGroupReaderIsPresent {
+			// there's a programmatic error. Something has gone wrong with tasklist instantiation
+			// don't block and redirect to the default group
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
+			tr.logger.Error("An isolation group buffer was misconfigured and couln't be found. Redirecting to default",
+				tag.IsolationGroup(group),
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+
+			select {
+			case <-tr.cancelCtx.Done():
+				// the task reader is shutting down
+				return true, true
+			case tr.taskBuffers[defaultTaskBufferIsolationGroup] <- taskInfo:
+				// task successfully rerouted to default tasklist
+				return false, true
+			default:
+				// couldn't redirect, loop and try again
+				return false, false
+			}
+		}
+
+		// if there is no poller in the isolation group or the isolation group is drained,
+		// we want to redistribute the tasks to other isolation groups in this case to drain
+		// the backlog
+		select {
+		case <-tr.cancelCtx.Done():
+			// the task reader is shutting down
+			return true, true
+		case tr.taskBuffers[group] <- taskInfo:
+			// successful redirect
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectCounter)
+			tr.logger.Warn("some tasks were redirected to another isolation group", tag.IsolationGroup(group),
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+			return false, true
+		default:
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
+			tr.logger.Error("some tasks could not be redirected to another isolation group as the buffer's already full", tag.IsolationGroup(group),
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+			// the task async buffers on the other isolation-group are already full, wait and retry
+		}
+	}
+	// this should never happen unless there is a bug - don't drop the task
+	tr.scope.IncCounter(metrics.BufferThrottlePerTaskListCounter)
+	tr.logger.Error("taskReader: unexpected error dispatching task", tag.Error(err))
+	runtime.Gosched()
+	return false, false
+}
```

Comment:
probably deserves a comment update at the very least, but I'm gonna pause on this and read a bit deeper in general:

since this changed from blocking to non-blocking, this is _essentially_ now no longer true - we'd expect it to occur when the destination buffer is full.

I'm not entirely sure what the consequences are of this though.  e.g. it'll try to re-dispatch to the original (now known to be old) group - is that fine?
it seems like probably yes, it'll just time out again and reach these lines again?  that probably serves as a throttler on the whole process to make sure it doesn't hot-loop, which would be good, as long as it doesn't incorrectly perturb the order or something.

Reply:
yeah, that does look safe to me.  waits on a ratelimit, tries to dispatch locally or forward, then fails.  timeout or "successfully dispatched" both seem like reasonable outcomes.

I'm not familiar with exactly what a forwarded task implies though.  I'm guessing no, but is there a chance this bounces around between hosts, wasting more resources than if it stayed here?

Reply:
Yeah, I think it's safe to retry / won't hotloop, in isolated cases it'll time out, if non-isolated case it'll block, in both I don't think it'll spin. 

The forwarding is to forward on to the parent tasklist in tasklist partitioning trees. I don't think it affects this (As far as I can tell)